### PR TITLE
bext: include code-intel extensions to native integration bundle

### PR DESCRIPTION
**Pull request by taras-yemets**
Closes https://github.com/sourcegraph/sourcegraph/issues/49380

Bundles code-intel extensions with the native integrations code.

## Test plan
- Ensure browser extension is working ([README](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@d1bb51b94cd90f960691fd5fde7cf4cb20fe5522/-/blob/client/browser/README.md?L70))
- Ensure Bitbucket integration is working:
  - `sg start` (ensure Bitbucket code host is configured)
  - `sg run bext`
  - log in to our Bitbucket instance https://bitbucket.sgdev.org
    - ensure Sourcegraph integration is enabled and it points to your local Sourcegraph instance ([README](https://sourcegraph.com/github.com/sourcegraph/bitbucket-server-plugin@97fc6d39f406b100359ee5e22cc3bee9d1453df8/-/blob/README.md?L33-38&view=code))
  - open repo on the Bitbucket instance and ensure that view repo/file on Sourcegraph and code-intel tooltips work as expected

## App preview:

- [Web](https://sg-web-taras-yemets-bext-fix-bitbucket.onrender.com/search)

Check out the [client app preview documentation](https://docs.sourcegraph.com/dev/how-to/client_pr_previews) to learn more.

### CODE REVIEW

File: client/browser/src/shared/platform/inlineExtensionsService.ts [hunk 1]
```diff
@@ -9,37 +9,52 @@ import extensions from '../../../code-intel-extensions.json'
...
-    // In a browser extension environment, we need to find the absolute ULR in the browser extension
-    // asssets directory.
+    /**
+     * In a browser extension environment, we need to find the absolute ULR in the browser extension assets directory.
+     * We assume extension bundles exist (e.g. were built using [build-inline-extensions](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@a2190db3ee4b7e61f02f821cd2e5b58fa9814540/-/blob/client/browser/scripts/build-inline-extensions.js) script).
+     */
     if (typeof window.browser !== 'undefined') {
         return {
-            manifestURL: browser.extension.getURL(manifestPath),
-            scriptURL: browser.extension.getURL(scriptPath),
```
Comment (taras-yemets) [inline client/browser/src/shared/platform/inlineExtensionsService.ts:23]: `extension.getURL()` has been [deprecated](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/extension/getURL).

File: client/browser/src/shared/platform/inlineExtensionsService.ts [hunk 2]
```diff
@@ -9,37 +9,52 @@ import extensions from '../../../code-intel-extensions.json'
...
+            manifestURL: browser.runtime.getURL(manifestPath),
+            scriptURL: browser.runtime.getURL(scriptPath),
```
Comment (taras-yemets) [inline client/browser/src/shared/platform/inlineExtensionsService.ts:27]: `extension.getURL()` has been [deprecated](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/extension/getURL).

File: client/browser/src/shared/platform/inlineExtensionsService.ts [hunk 3]
```diff
@@ -9,37 +9,52 @@ import extensions from '../../../code-intel-extensions.json'
...
         }
     }
-    // If SOURCEGRAPH_ASSETS_URL is defined (happening for e.g. GitLab native extenisons), we can
-    // use this URL as the root for bundled assets.
+
+    /**
+     * In a native integration environment if SOURCEGRAPH_ASSETS_URL is defined (e.g. GitLab, see [sourcegraph/index.js](https://sourcegraph.com/gitlab.com/gitlab-org/gitlab@a99a4b33543b01d05abd5fcd76b05298ae614fe4/-/blob/app/assets/javascripts/sourcegraph/index.js)),
+     * we can use this URL as the root for bundled assets.
+     */
```
Comment (taras-yemets) [inline client/browser/src/shared/platform/inlineExtensionsService.ts:34]: @philipp-spiess, you've been working on a GitLab integration fix:
- https://github.com/sourcegraph/sourcegraph/pull/42106

If we go with the approach suggested by this PR (include code-intel extensions into integration bundle), do we still need `SOURCEGRAPH_ASSETS_URL` code path?

**Review by philipp-spiess — COMMENTED** [review]
I don't understand this fix. This change suggests that we no longer use the bundled version of the extension now for native extensions but instead use the `sourcegraphURL` which means that it downloads the extension code from the sourcegraph instance? 🤔 

Isn't that the opposite of what we want?

File: client/browser/src/shared/platform/inlineExtensionsService.ts (hunk 3 above)
Comment (philipp-spiess) [reply client/browser/src/shared/platform/inlineExtensionsService.ts:34]: I’m pretty sure GitLab wants to continue serving the code from their own origin and not a potential third-party (`sourcegraphURL` is the sourcegraph server URL, no?)

**Comment by philipp-spiess** [comment]
I’m not sure if this fix works on a production server. I think it only works for you because we do this: https://github.com/sourcegraph/sourcegraph/pull/49396/files#diff-4d4339e928b07ffc5c5d2009d6e52e5bed181767b4bec5bbe18558a1a0bb627eR129-R130

which means that these assets will be served by your local dev setup but not from the production bundle.

I think what we need to fix this is doing exactly what GitLab is doing here:

https://sourcegraph.com/gitlab.com/gitlab-org/gitlab@a99a4b33543b01d05abd5fcd76b05298ae614fe4/-/blob/app/assets/javascripts/sourcegraph/index.js?L19-23

We need to:

- detect where the script is being loaded from
- set `SOURCEGRAPH_ASSETS_URL` based on that value

**Comment by philipp-spiess** [comment]
Oh, right, in the case of the third-party extension, we do serve the native extension from the sourcegraph URL. So I can see why it works in this case. 

In this case the asset URL is hosed on the same URL as the Sourcegraph server

**Comment by taras-yemets** [comment]
> Oh, right, in the case of the third-party extension, we do serve the native extension from the sourcegraph URL. So I can see why it works in this case.
> 
> In this case the asset URL is hosed on the same URL as the Sourcegraph server

X-posting some findings from the [Slack thread](https://sourcegraph.slack.com/archives/C04931KQVRC/p1678813965649589?thread_ts=1678345162.896039&cid=C04931KQVRC) for context:
> Sourcegraph native integration for Bitbucket Server [uses](https://sourcegraph.com/github.com/sourcegraph/bitbucket-server-plugin@97fc6d39f406b100359ee5e22cc3bee9d1453df8/-/blob/src/main/resources/js/sourcegraph-bitbucket.js?L23) extension/scripts/integration.bundle.js on the connected Sourcegraph instance to add code-intel features to the code host. So the native integration code is tied to the Sourcegraph version.
...
But as was previously mentioned, Sourcegraph native integration also doesn’t work for our Bitbucket Server instance (https://bitbucket.sgdev.org/) even if connected to the “freshest” Sourcegraph instance (s2 or local one built on the current main). Links to view the repo and file on Sourcegraph are rendered, but the code-intel doesn’t work.
When debugging the issue, I ran into the "Can not construct extension URLs" error, meaning that the [code](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@9e2473e0e060d9542fc8315241886bd84a30a932/-/blob/client/browser/src/shared/platform/inlineExtensionsService.ts?L35)-intel bundles were not loaded (code).
Once we find the fix for this (and maybe subsequent) issue, it can be included in the next Sourcegraph release (or even backported to 5.0).

File: client/browser/scripts/tasks.ts [hunk 4]
```diff
@@ -125,6 +125,7 @@ export function copyIntegrationAssets(): void {
     shelljs.cp('build/dist/css/style.bundle.css', 'build/integration/css')
     shelljs.cp('build/dist/css/inject.bundle.css', 'build/integration/css')
     shelljs.cp('src/native-integration/extensionHostFrame.html', 'build/integration')
+    copyInlineExtensions('build/integration')
```
Comment (taras-yemets) [inline client/browser/scripts/tasks.ts:128]: Similar to [how we build](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@0a7bfc6c389b71484615e788a62b9ef262823726/-/blob/client/browser/scripts/build-npm.ts?L47) for the npm package used on GitLab.

File: client/browser/src/shared/platform/inlineExtensionsService.ts [hunk 5]
```diff
@@ -9,37 +9,45 @@ import extensions from '../../../code-intel-extensions.json'
...
+     * • GitLab native integration and its assets are hosted on a corresponding GitLab instance (see {@link https://sourcegraph.com/gitlab.com/gitlab-org/gitlab@a99a4b33543b01d05abd5fcd76b05298ae614fe4/-/blob/app/assets/javascripts/sourcegraph/index.js|sourcegraph/index.js} and {@link https://github.com/sourcegraph/sourcegraph/pull/42106|pull/42106})
+     *
+     * • Bitbucket Server integration and its assets are hosted on a connected Sourcegraph instance (see {@link https://sourcegraph.com/github.com/sourcegraph/bitbucket-server-plugin@97fc6d39f406b100359ee5e22cc3bee9d1453df8/-/blob/src/main/resources/js/sourcegraph-bitbucket.js|sourcegraph-bitbucket.js} and {@link copyIntegrationAssets})
+     */
+    if (assetsURL) {
         return {
-            manifestURL: new URL(manifestPath, window.SOURCEGRAPH_ASSETS_URL).toString(),
-            scriptURL: new URL(scriptPath, window.SOURCEGRAPH_ASSETS_URL).toString(),
+            manifestURL: new URL(manifestPath, assetsURL).toString(),
+            scriptURL: new URL(scriptPath, assetsURL).toString(),
```
Comment (taras-yemets) [inline client/browser/src/shared/platform/inlineExtensionsService.ts:39]: Use either `window.SOURCEGRAPH_ASSETS_URL` or `sourcegraphURL` (computed in [`getAssetsURL`](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@31c61ada692445360a3e60a85042f714468a4bc9/-/blob/client/browser/src/shared/util/context.tsx?L19-32)).

File: client/browser/src/shared/platform/inlineExtensionsService.ts [hunk 6]
```diff
@@ -9,37 +9,45 @@ import extensions from '../../../code-intel-extensions.json'
...
+function getURLsForInlineExtension(extensionID: string, assetsURL: string): { manifestURL: string; scriptURL: string } {
     const kebabCaseExtensionID = extensionID.replace(/^sourcegraph\//, 'sourcegraph-')
-
     const manifestPath = `extensions/${kebabCaseExtensionID}/package.json`
     const scriptPath = `extensions/${kebabCaseExtensionID}/extension.js`
 
-    // In a browser extension environment, we need to find the absolute ULR in the browser extension
-    // asssets directory.
+    /**
+     * In a browser extension environment, we need to find the absolute ULR in the browser extension assets directory.
```
Comment (marekweb) [inline client/browser/src/shared/platform/inlineExtensionsService.ts:18]:
```suggestion
     * In a browser extension environment, we need to find the absolute URL in the browser extension assets directory.
```

**Pull request merged**